markdown/postprocessors.py

```diff
@@ -75,9 +75,18 @@ def run(self, text):
                     self.md.htmlStash.get_placeholder(i))] = html
             replacements[self.md.htmlStash.get_placeholder(i)] = html
 
+        def substitute_match(m):
+            key = m.group(0)
+
+            if key not in replacements:
+                return f'<p>{ replacements[key[3:-4]] }</p>'
```

Comment:
I figured out what I don't like about this. If `key[3:-4]` is also not in `replacements`, then this will raise an error. Also, this assumes that `key` includes `<p>` and `</p>`. However it might not. Seems to me that the correct check would be:

```python
if key not in replacements and key.startswith('<p>') and key.endswith('</p>') and key[3:-4] in replacements:
```

I wonder if perhaps there could be some other scenarios which would require slightly different behavior. For example, we might need an `elif` statement. However, until we find an edge case that fits, we'll have to wait to see if they ever get reported. It's also possible that such cases will never occur.

Reply:
I suppose `key.startswith('<p>') and key.endswith('</p>')` doesn't need to be in the if statement. `key[3:-4] in replacements` will only evaluate to `True` when that is the case anyway.

Reply:
I'm thinking this should be:

```python
if key not in replacements:
    if key[3:-4] in replacements:
        return f'<p>{ replacements[key[3:-4]] }</p>'
    else:
        return key
```

We just need a test case for the final `else`.

Reply:
Here is the test:

```python
def test_placeholder_in_source(self):
    # This should never occur, but third party extensions could create weird edge cases.
    md = markdown.Markdown()
    # Ensure there is an htmlstash so relevant code (nested in `if replacements`) is run.
    md.htmlStash.store('foo')
    # Run with a placeholder which is not in the stash
    placeholder = md.htmlStash.get_placeholder(md.htmlStash.html_counter + 1)
    result = md.postprocessors['raw_html'].run(placeholder)
    self.assertEqual(placeholder, result)
```

Without my proposed change, this test raises a `KeyError`. With my proposed change and this test, we get 100% test coverage of the diff.

Reply:
If I understand correctly, the case of concern is if both `key` and `key[3:-4]` are not in `replacements`, which implies that the regexp search found some sequence that looks like `'\x02wzxhzdk:1234\x03'`(for any digits substituting 1234) that is not a place-holder for a pending substitution, correct?

The `replacements` ordered dict is constructed immediately prior in lines [70-76](https://github.com/Python-Markdown/markdown/blob/48720f1f79480307d384bbfd8e46765ca1430439/markdown/postprocessors.py#L70), so the code is tightly coupled without assumptions from other files or piece of the codebase. All substitutions in the htmlStash get an a key in the `replacement` dict without the `<p>` tags, but some also get keys featuring the `<p>` tags (and with higher priority, since this is an ordered dict).

Based on how `replacements` is constructed as of now, `key not in replacements` is `True` only when `key` has the `<p>` tags. A subsequent `key[3:-4] in replacements` will always evaluate to `True`.

Let me know if this answers questions, or if there are still concerns you have about robustness.

Reply:
First of all, if you pass text into Markdown which would match, by the time we get to the postporcessor that is no longer a concern. However, a third-party extension could insert text into the document which would match, but have no corresponding key. Therefore, we need to account for that. My fix does that.

Reply:
Ah. I had not delved into the interaction of extensions with this code. Excellent. Thank you for your help and for accepting the overall changes.